pytorch_lightning/trainer/connectors/data_connector.py

```diff
@@ -60,12 +74,43 @@ def on_trainer_init(
         self.trainer.reload_dataloaders_every_n_epochs = reload_dataloaders_every_n_epochs
         self.trainer._is_data_prepared = False
 
-    def get_profiled_train_dataloader(self, train_dataloader) -> Iterable:
-        self.data_fetcher = DataFetcher()
-        self.data_fetcher.setup(train_dataloader)
-        prefetcher_iter = iter(self.data_fetcher)
-        profiled_dl = self.trainer.profiler.profile_iterable(enumerate(prefetcher_iter), "get_train_batch")
-        return profiled_dl
+    def _check_training_step_requires_dataloader_iter(self) -> bool:
+        training_step_fx = getattr(self.trainer.lightning_module, "training_step")
+        contains_datalaoder_iter = is_param_in_hook_signature(training_step_fx, "dataloader_iter", explicit=True)
+        use_manual_optimization = not self.trainer.lightning_module.automatic_optimization
+        if contains_datalaoder_iter and not use_manual_optimization:
+            raise MisconfigurationException(
+                "When the `training_step` has a `dataloader_idx` argument, you should set "
+                "`automatic_optimization=False` within your LightningModule. "
+            )
+
+        return contains_datalaoder_iter and use_manual_optimization and self.trainer.training
+
+    def _select_data_fetcher(self) -> AbstractDataFetcher:
+        if self._check_training_step_requires_dataloader_iter():
+            log.warning(
+                "Found `dataloader_iter` argument in the `training_step`. Note that the support for "
+                "this signature is experimental and the behavior may subject to change."
+            )
+            return DataLoaderIterDataFetcher()
+        elif self.trainer.training_type_plugin.on_gpu and os.getenv("PL_INTER_BATCH_PARALLELISM", "0") == "1":
+            # experimental feature
+            return InterBatchParallelismDataFetcher()
+        else:
+            return DataFetcher()
+
+    def get_profiled_dataloader(self, dataloader: Iterable, dataloader_idx: int = 0) -> Iterable:
```

Comment:
the naming is confusing: this accepts a dataloader but returns the iterable. IMO the loop should call `iter(data_fetcher)` and this could simply be `get_datafetcher` 

it's somewhat confusing that this accepts a dataloader and returns an iterator

Reply:
Fine :)